### PR TITLE
feat(image-viewer): create dtImageViewer

### PR DESCRIPTION
**Pull request by AlexLamDialpad**
# PR Title

[DP-67072](https://dialpad.atlassian.net/browse/DP-67072)
[DT-994](https://dialpad.atlassian.net/browse/DT-994)
## :hammer_and_wrench: Type Of Change

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Adds an image viewer DT component to be used in DP2.0

## :bulb: Context

We want to create a well defined image viewer to be used in Dialpad 2.0 and throughout the rest of Dialpad.
It should support these features:
- Previewing the image in a smaller resolution
- Opening into a modal that features the full resolution image (Up until 80% of the screen size)
- The image should scale accordingly
- The image should allow the user to click out of, press escape, or close out of the modal
Create an image viewer to be used in the Dialpad product.

## :pencil: Checklist

- [x] I have reviewed my changes
- [x] I have added tests
- [x] I have added all relevant documentation
- [x] I have validated components with a screen reader
- [x] I have validated components keyboard navigation
- [x] I have considered the performance impact of my change
- [ ] I have checked that my change did not significantly increase bundle size
- [x] I am exporting any new components or constants in the index.js in the component directory
- [x] I am exporting any new components or constants in the index.js in the root

## :camera: Screenshots / GIFs

image
image
image

## :link: Sources

[DP-67072]: https://dialpad.atlassian.net/browse/DP-67072?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DT-994]: https://dialpad.atlassian.net/browse/DT-994?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by AlexLamDialpad** [comment]
@braddialpad Having trouble with with this piece of code in the test
```

    focusAfterOpen () {
      this.$refs.closeImage.$el.focus();
    },
```

image

Any ideas?

**Comment by braddialpad** [comment]
> @braddialpad Having trouble with with this piece of code in the test

Hmm I've definitely had a bunch of issues with focus in jsdom... Looking into it.

**Comment by braddialpad** [comment]
> @braddialpad Having trouble with with this piece of code in the test

Guessing it's transition related. The error is happening in `focusAfterOpen` and `this.$refs.closeImage` is not existing at the time it was called, as if it's not yet "completed" the transition. 

Check out popover.test.js where we have:

```
const transitionStub = () => ({
  render: function (h) {
    return this.$options._renderChildren;
  },
});

stubs: {
  // this gets around transition async problems. See https://v1.test-utils.vuejs.org/guides/common-tips.html
  transition: transitionStub(),
},
```

and try this in your test here (really annoying I know). Might also need:

```
global.requestAnimationFrame = sinon.spy();
global.cancelAnimationFrame = sinon.spy();
```

File: components/image_viewer/image_viewer.vue
Comment (AlexLamDialpad) [inline components/image_viewer/image_viewer.vue]: Should we pass in an aria-label here to specify that this button opens the image into a modal?

File: components/image_viewer/image_viewer_constants.js
Comment (AlexLamDialpad) [inline components/image_viewer/image_viewer_constants.js:1]: Should I remove this file if I didn't need it?

**Comment by AlexLamDialpad** [comment]
TODO before merge: Replace the images in the PR

I asked Bradley from design to see if there are stock images we could use for this.

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [reply components/image_viewer/image_viewer.vue]: Yes

File: components/image_viewer/image_viewer_constants.js
Comment (braddialpad) [reply components/image_viewer/image_viewer_constants.js:1]: Sure

File: components/image_viewer/test.jpg
Comment (iropolo) [inline components/image_viewer/test.jpg:1]: Is it necessary to have this 3.12MB image to test?

File: components/image_viewer/image_viewer.vue
Comment (iropolo) [inline components/image_viewer/image_viewer.vue]: If it is not necessary, I recommend that it be removed

File: components/image_viewer/test.jpg
Comment (iropolo) [reply components/image_viewer/test.jpg:1]: -----
https://github.com/dialpad/dialtone-vue/pull/834#issuecomment-1476572448 👍🏼

**Comment by iropolo** [comment]
Strange behaviour If you open the image then click it, the X button disappear, then it appear again when you hover the gray area.

Im wondering if it necessary to have this "show & hide" behaviour the close button.
If its a modal Im guessing we should always show the close button.

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [inline components/image_viewer/image_viewer.vue]: `w-wm332` this a typo?

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [inline components/image_viewer/image_viewer.vue]: Use portal (teleport in vue3) to render this at the body.

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [inline components/image_viewer/image_viewer.vue]: Do we actually need to use d-modal__dialog here? We can if it makes sense but I think there may be a lot of stuff in here we don't need to use in the case of this image viewer. for reference here's the class:

```css
.d-modal__dialog {
  position: relative;
  z-index: var(--zi-hide);
  display: flex;
  flex-direction: column;
  gap: var(--space-500);
  box-sizing: border-box;
  width: 100%;
  max-width: calc(var(--size-300) * 157); // 628
  max-height: 100%;
  padding: var(--modal-dialog-padding); // 32
  overflow-y: auto;
  color: var(--modal-dialog-color-text);
  font-size: var(--fs-200);
  line-height: var(--lh-400);
  background-color: var(--modal-dialog-color-background);
  background-clip: padding-box;
  border: var(--size-100) solid var(--modal-dialog-color-border);
  border-radius: var(--size-500);
  box-shadow: var(--modal-dialog-shadow);
  transform: translate3d(0, 30%, 0) scale3d(0.75, 0.75, 0.75);
  backface-visibility: hidden;
  visibility: hidden;
  opacity: 0;
  will-change: visibility, z-index, opacity, transform;
}
```

Might make sense to create a dedicated class for the image viewer, also would prevent accidentally breaking things on the image viewer when making style changes to the modal dialog.

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [reply components/image_viewer/image_viewer.vue]: Yeah you can remove this

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [inline components/image_viewer/image_viewer.vue]: Call `imageSrc` just for consistency with some of our other components that do this.

**Review by braddialpad — CHANGES_REQUESTED** [review]
Lookin pretty good so far!

File: components/image_viewer/image_viewer.vue
Comment (braddialpad) [inline components/image_viewer/image_viewer.vue:82]: Any way to set the size of the thumbnail? This seems like something that might need to be commonly done.

Comment (AlexLamDialpad) [reply components/image_viewer/image_viewer.vue:82]: Added imageButtonClass

**Comment by AlexLamDialpad** [comment]
> TODO before merge: Replace the images in the PR
> 
> I asked Bradley from design to see if there are stock images we could use for this.

The image is taken from [unsplash](https://unsplash.com/)
image

Gifs I'm not really clear on whether we need special commercial usage gifs? Seems very grey zone.

Maybe we should get some marketing pictures and gifs instead?

**Comment by AlexLamDialpad** [comment]
> Strange behaviour If you open the image then click it, the X button disappear, then it appear again when you hover the gray area.
> 
> Im wondering if it necessary to have this "show & hide" behaviour the close button. If its a modal Im guessing we should always show the close button.

It matches the current behaviour and it's spec'ed out in the figma design 🤷

**Comment by braddialpad** [comment]
> Gifs I'm not really clear on whether we need special commercial usage gifs? Seems very grey zone.

I doubt you have to worry about this, we're just using these for examples right?

**Review by braddialpad — COMMENTED** [review]
This is looking pretty good to me, but I know design still wants to have a look at it so I'll hold off a bit before  approving.

**Comment by bhawkins-dialpad** [comment]
Thanks @AlexLamDialpad and @braddialpad I think it looks pretty good. The showing/hiding of the close button is intentional. In the future we wanted to add more detail about the image in this header area (file name, author, date, etc...). This would allow the user to focus just on the image and hide that extra information.

Would it be possible to include some the transitions on the prototype to smooth out the experience a little bit? 
- Image enlarges from the conversation view thumbnail with a slight spring animation and reduces in size when closing the enlarge view.  
- The close button fades and moves into view.

Happy to talk about these transitions further to see what is possible.

[Figma prototype](https://www.figma.com/proto/vtp3rfT44p9yjur7Mxb3HH/Messaging---Enlarge-Image?page-id=635%3A20989&node-id=659-39664&viewport=2671%2C763%2C0.37&scaling=min-zoom&starting-point-node-id=659%3A39664&show-proto-sidebar=1)

https://user-images.githubusercontent.com/84408065/227541438-53f3d2f3-b7c4-4bbc-9759-ff952a4d69c6.mov

**Comment by AlexLamDialpad** [comment]
> Thanks @AlexLamDialpad and @braddialpad I think it looks pretty good. The showing/hiding of the close button is intentional. In the future we wanted to add more detail about the image in this header area (file name, author, date, etc...). This would allow the user to focus just on the image and hide that extra information.
> 
> Would it be possible to include some the transitions on the prototype to smooth out the experience a little bit?
> 
> * Image enlarges from the conversation view thumbnail with a slight spring animation and reduces in size when closing the enlarge view.
> * The close button fades and moves into view.
> 
> Happy to talk about these transitions further to see what is possible.
> 
> [Figma prototype](https://www.figma.com/proto/vtp3rfT44p9yjur7Mxb3HH/Messaging---Enlarge-Image?page-id=635%3A20989&node-id=659-39664&viewport=2671%2C763%2C0.37&scaling=min-zoom&starting-point-node-id=659%3A39664&show-proto-sidebar=1)

@bhawkins-dialpad 
I added the transition on the button to fade in and out as the cursor moves in and out of the window.
Adding in a transition into the actual image opening I've had far less luck. 

Are we okay with making that a future upgrade?

**Comment by bhawkins-dialpad** [comment]
@AlexLamDialpad Thanks! The close transitions look better. I think its ok to hold off on the greater image transition/animation. There probably needs to be more thought put into our motion strategy and make sure thats the right treatment and maybe we add it later. cc @francisrupert

**Review by braddialpad — APPROVED** [review]

**Pull request merged**
